lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1797,6 +1804,64 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
   KPE->resolveComponents(TC.Context, components);
 }
 
+Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
+  // If constructor call is expected to produce an optional let's not attempt
+  // this optimization because literal initializers aren't failable.
+  if (!ExprStack.empty()) {
+    auto *parent = ExprStack.back();
+    if (isa<BindOptionalExpr>(parent) || isa<ForceValueExpr>(parent))
+      return nullptr;
+  }
```

Comment:
I was just trying to avoid breaking existing code like https://github.com/apple/swift/pull/17860/files#diff-de58b3c2866a24533bc39d1bfd40fc3cR276 but maybe I shouldn't do it at all then?

Reply:
Alright, this check is hidden under `!isSwiftVersionAtLeast(5)` now.